x-pack/filebeat/input/awss3/s3_input.go

```diff
@@ -188,11 +207,17 @@ func (in *s3PollerInput) workerLoop(ctx context.Context, workChan <-chan state)
 			err := in.states.AddState(state)
 			if err != nil {
 				in.log.Errorf("saving completed object state: %v", err.Error())
+				in.status.UpdateStatus(status.Degraded, fmt.Sprintf("Failure saving completed object state: %s", err.Error()))
 			}
 
 			// Metrics
 			in.metrics.s3ObjectsAckedTotal.Inc()
 		})
+		// if we go from loop iteration failure to success, we should attempt to re-report recovered status
+		// the purpose of this block is to avoid unnecessary synchronization between workers during status update
+		if state.Stored {
+			in.status.UpdateStatus(status.Running, "Input is running")
+		}
```

Comment:
IMO this can be done within the acknowledgement callback function (see `acks.Add` above). This is where we increment the `s3ObjectsAckedTotal` which correlates to successfully processing of an S3 file.